testonly/hammer/record.go

```diff
@@ -75,44 +75,62 @@ func (g *sharedState) proposeLeaves(rev uint64, leaves []*trillian.MapLeaf) erro
 	return nil
 }
 
-// TODO(mhutchinson): Make this way more tolerant so it accepts older SMRs and
-// checks they are equivalent to previously seen versions if applicable.
 func (g *sharedState) advertiseSMR(smr types.MapRootV1) error {
-	g.mu.Lock()
-	defer g.mu.Unlock()
+	update, err := func() (bool, error) {
+		g.mu.RLock()
+		defer g.mu.RUnlock()
+
+		if g.smrs[0] == nil || g.smrs[0].Revision < smr.Revision {
+			return true, nil
+		}
 
-	var prevRev uint64 // The last revision committed to contents.
-	if g.smrs[0] != nil {
-		if g.smrs[0].Revision > smr.Revision {
-			return fmt.Errorf("stale root: got revision %d, already had %d", smr.Revision, g.smrs[0].Revision)
-		} else if g.smrs[0].Revision == smr.Revision {
-			if !reflect.DeepEqual(g.smrs[0], &smr) {
-				return fmt.Errorf("SMR mismatch at revision %d: had %+v, got %+v", smr.Revision, g.smrs[0], smr)
+		for i := 0; i < smrCount && g.smrs[i] != nil && g.smrs[i].Revision >= smr.Revision; i++ {
+			if g.smrs[0].Revision == smr.Revision {
+				if !reflect.DeepEqual(g.smrs[i], &smr) {
+					return false, fmt.Errorf("SMR mismatch at revision %d: had %+v, got %+v", smr.Revision, g.smrs[0], smr)
+				}
 			}
-			// Roots are equal, so no need to push on the same root twice
-			return nil
 		}
-		prevRev = g.smrs[0].Revision
+		return false, nil
+	}()
+	if err != nil {
+		return err
 	}
 
-	glog.V(2).Infof("adding new SMR: %+v", smr)
-	// Shuffle earlier SMRs along.
-	for i := smrCount - 1; i > 0; i-- {
-		g.smrs[i] = g.smrs[i-1]
-	}
+	if update {
+		g.mu.Lock()
+		defer g.mu.Unlock()
+
+		// update is a suggestion. We need to confirm the condition holds now we have write lock.
+		if g.smrs[0] != nil && g.smrs[0].Revision >= smr.Revision {
```

Comment:
This check could be avoided if the lock wasn't released. I wonder if this `if update` case can be just inlined to line 83. And the you'd also not need the temporary function.

Reply:
This would involve trying to upgrade the read lock to a write lock, or making all operations take a write lock. I'm not particularly excited by that. Did you miss the subtle RLock vs Lock, or am I missing something that makes this a good idea?

Reply:
I see, yeah, I missed that bit. SGTM.